server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -255,6 +271,24 @@ public Index concreteSingleIndex(ClusterState state, IndicesRequest request) {
         return indices[0];
     }
 
+    /**
+     * Utility method that allows to resolve an index expression to its corresponding single write index.
+     *
+     * @param state             the cluster state containing all the data to resolve to expression to a concrete index
+     * @param request           The request that defines how the an alias or an index need to be resolved to a concrete index
+     *                          and the expression that can be resolved to an alias or an index name.
+     * @throws IllegalArgumentException if the index resolution does not lead to an index, or leads to more than one index
+     * @return the write index obtained as a result of the index resolution
+     */
+    public Index concreteWriteIndex(ClusterState state, IndicesRequest request) {
+        String indexExpression = request.indices() != null && request.indices().length > 0 ? request.indices()[0] : null;
+        Context context = new Context(state, request.indicesOptions(), false, true);
+        Index[] indices = concreteIndices(context, indexExpression);
+        // concreteIndices will throw its own exception when checking for a write index. Assert here for good measure.
+        assert indices.length == 1 : "The index/alias and options provided did not point to a write-index";
```

Comment:
I think this needs to be stronger - the expression resolvers can in theory give you multiple expressions / indices. This needs to be a hard exception (and we should test it)

Reply:
you're right, need more tests for multiple index expressions

Reply:
after looking at this for a second... looks like the intent of this method is to be called against DocWriteRequests, which, whether they actually have multiple expressions or not, the above lines filter for the first index expression. So conreteIndices will never be called with multiple index expressions, even if the request did have them. That is why I could never reach this situation. With write index being requested, concreteIndices will throw an exception if nothing is found, otherwise it will add the one write index and continue on

Reply:
Even though DocWriteRequest makes sense here, it is not up to the DocWriteRequest to specify that it would

1. have only one index expression in indices()
2. use IndicesOptions.strictSingleIndexNoExpandForbidClosed()

So, leaving it as IndicesRequest is just as good it seems